### ISSUE TITLE
Derived variable needs different input dataset for CMIP5/CMIP6

### ISSUE THREAD

**Issue by ledm** [ESMValGroup/ESMValCore#310]
The AMOC derived variable needs to use different data depending on whether we're looking at CMIP5 or CMIP6 data. This is because the input variable that was used in CMIP5 (**msftmyz**) has been superseded by a slightly different variable in CMIP6 (**msftyz**). 

Is there a way to derive a variable such that it checks for the presence/absence of the two variables **msftmyz** or **msftyz** then uses the best one? It shouldn't fail if it doesn't find one dataset, provided that it finds the other.

(The difference between the two is that msftmyz is defined in the meridional direction and msftyz is defined in the Y direction.)

I like to keep you on your toes! Enjoy Bavaria this week!

**Comment by mattiarighi** [comment]
I think @axel-lauer had a similar issue some time ago, but I can't remember the details.

**Comment by zklaus** [comment]
@debe-kevin also just stumbled upon this.

There is a facility in the cmor checker for variables that have changed name from CMIP5 to CMIP6, namely the mapping in https://github.com/ESMValGroup/ESMValCore/blob/20296fc0ec576128a31a670e9fc0bda4e3483138/esmvalcore/cmor/table.py#L73

@ledm I am not completely clear on the move: there seem to be two variables that can be considered to be the new versions of `msftmyz` in CMIP6: `msftmz` and `msftyz`. My intuition is that actually `msftmz` is closer to `msftmyz` in spirit. What do you think?

**Comment by schlunma** [comment]
Is this fixed by #230 @valeriupredoi? Did not try it, just guessing from reading the source code :grimacing:

**Comment by ledm** [comment]
I've been looking at what data is available and I've seen that the follow variable have CMIP6 model data in the historical r1* ensemble member:

- `msftmyz` : GISS-E2-1-G, GISS-E2-1-G
- `msftmz` : CanESM5, CESM2, CESM2-WACCM, SAM0-UNICON
- `msftyz` : CNRM-CM6-1, CNRM-CM6-1, CNRM-ESM2-1, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, EC-Earth3-Veg, IPSL-CM6A-LR, HadGEM3-GC31-LL, UKESM1-0-LL

It looks like the CMIP6 model have done a pick-n-mix with their favourite streamfunction variable. Perhaps it would be better to use a preferential cascade to determine the AMOC (ie, first try to use `msftmyz`, then try `msftmz` then try `msftyz`?)

A valid scientific question arises: Is there a difference between the meridional overturning circulation and the Y-direction overturning circulation? @tillku @shaynemcgregor, any thoughts?

**Comment by zklaus** [comment]
`msftmyz` is wrong. This variable has been retired in december 2017 in favor of the other two. These models simply need fixes.

The difference between the other two is described in the data request, more precisely in the associated `standard_names`. Quoting from the data request for[msftyz](http://clipc-services.ceda.ac.uk/dreq/u/baa5a662-e5dd-11e5-8482-ac72891c3257.html):

> Function of Y, Z, basin. For a model with a cartesian latxlon grid, this is the same as the "Ocean Meridional Overturning Mass Streamfunction" (msftmz) and should in this case be omitted.

and for [msftmz](http://clipc-services.ceda.ac.uk/dreq/u/baa59d48-e5dd-11e5-8482-ac72891c3257.html)

> Function of latitude, Z, basin. For a model with a cartesian latxlon grid, this is the same as the "Ocean Y Overturning Mass Streamfunction" (msftyz), which should in this case be omitted. For other models, this transport should be approximated as the transport across zig-zag paths corresponding to latitudes with spacing between latitudes appropriate to the model's resolution.

**Comment by ledm** [comment]
Okay, so now we know which variable to look at - how do we implement this in a derived variable? 

At the moment, I've tried:
```
    # Required variables
    required = [{'short_name': 'msftmyz', 'project': 'CMIP5'},
                {'short_name': 'msftyz', 'project': 'CMIP6'}]
```
but that doesn't work as it still looks for both fields. 

Do we need to do a re-write to the derivation module?

**Comment by schlunma** [comment]
> Okay, so now we know which variable to look at - how do we implement this in a derived variable?
> 
> At the moment, I've tried:
> 
> ```
>     # Required variables
>     required = [{'short_name': 'msftmyz', 'project': 'CMIP5'},
>                 {'short_name': 'msftyz', 'project': 'CMIP6'}]
> ```
> 
> but that doesn't work as it still looks for both fields.
> 
> Do we need to do a re-write to the derivation module?

In #230 `required` is going to be a function dependent on `project`, i.e. it will be possible to do

```python
@staticmethod
def required(project):
    if project == 'CMIP5':
        ....
    else:
        ....
```

**Comment by ledm** [comment]
@zklaus, do that facility work for derived variables as well? It hasn't work for me in this case yet.

**Comment by zklaus** [comment]
What do you mean? Neither `msftmyz` in CMIP5 nor the two CMIP6 variables are derived, are they?

**Comment by zklaus** [comment]
Ah sorry, of course you are speaking about the amoc derived variable. Forget what I said. Let me look into that (and by that I mean pestering @schlunma, of course).

**Comment by zklaus** [comment]
@ledm you can have a look at [the derivation of `sispeed`](https://github.com/ESMValGroup/ESMValCore/blob/development/esmvalcore/preprocessor/_derive/sispeed.py) for an example.
Basically you then use the CMIP5 name in the required attribute and for the CMIP6 variables it will use the new name instead. This should solve the issue for the `msftmz` models, which is what I would start with.

Then to tackle those providing only `msftyz` maybe we can add a derivation of `msftmz`. Whether to do this inaccurately from `msftyz` (probably still good enough) or do something more involved to get it from a 3d field can be discussed. So all in all the required block for `amoc` then should look like
```
required = [{'short_name': 'msftmyz', 'derive': True}]
```
Sound good?

**Comment by ledm** [comment]
Sorry, I just don't get it. How is `sispeed` similar? There doesn't appear to be a cmip5/cmip6 naming difference there.

Also, it looks like PR #230 should resolve this issue right, @schlunma? If so, it might be better to wait until that PR is accepted...

**Comment by schlunma** [comment]
> Also, it looks like PR #230 should resolve this issue right, @schlunma? If so, it might be better to wait until that PR is accepted...

I thinks so, yes (I didn't test it for this particular case, though).

**Comment by zklaus** [comment]
> Sorry, I just don't get it. How is `sispeed` similar? There doesn't appear to be a cmip5/cmip6 naming difference there.

There totally is: The variables were called `uis` and `vis` in CMIP5, but are now named `siu` and `siv` in CMIP6.

**Comment by valeriupredoi** [comment]
USI as in the :gun: ? :grin:
#230 is able to fix the problem since `project` is now a derived required func optional attribute. But the question is more pedantic here: mixing variables from different CMIP projects is probably not the most statistically correct procedure: CMIP5 and CMIP6 have different populations obtained from different models with different initial conditions, forcings etc - in short, have different distributions that are not drawn from the same master distribution, I don't think it's an apples to apples situation :beer:

**Comment by zklaus** [comment]
Well, what should be compared is a decision for scientists to make, no? Our job here is to make it technically possible.

As for `amoc` specifically: The CMIP5 vs CMIP6 part is already solved. This just needs an addition to the mapping dictionary [as described above](https://github.com/ESMValGroup/ESMValCore/issues/310#issuecomment-543093719).

The question of `msftmz` vs `msftyz` is a bit more tricky. Iff one is convinced that `msftyz` can be used as is, then #230 can help with that by having both `msftmyz` and `msftyz` as optional variables in the requirements, using whichever of the two is available to do the computation.

This has nothing to do with CMIP5 vs CMIP6, but maybe is the real question here anyway?

**Comment by valeriupredoi** [comment]
> Well, what should be compared is a decision for scientists to make, no? Our job here is to make it technically possible.

yeps, just my two pennies about the statistical inconsistency, man

The only problem I foresee for this sort of case is that the two variables (one from CMIP5 and the other from CMIP6) are on different grids so the regridding (which should be done before the derived variable computation) should be done on a certain grid - which one? :beer:

**Comment by zklaus** [comment]
Let's stay on topic here. I think the issue as described in the title and the OP is solved: Add an entry to the dictionary and use the CMIP5 name. Let's close this issue. @ledm do you agree?

If there is another issue, we can open another issue.

As for @valeriupredoi's comment about the grids: I really don't see why the derived variable should not live on different grids in general. Again this seems to have nothing to do with CMIP5 vs CMIP6 since different CMIP6 models also come on different grids with resulting derived variables living potentially on different grids as well.
In this case of the stream function it is particularly simple since the only lateral dimension is latitude.

**Comment by valeriupredoi** [comment]
aye, agreed with @zklaus let's not dilute the matter 9which is solved by #230 ) :beer:

**Comment by zklaus** [comment]
#230 is great. It will help with many things. This issue, however, is unrelated and can just be solved/closed with the current infrastructure.

**Comment by ledm** [comment]
I'm finding a whole series of problems with CMIP6 models data for msftyz. No single CMIP6 dataset can be read with no fixes and some of the changes required are rather extensive.

**Comment by zklaus** [comment]
@ledm, since these issues seem not to be related with the topic of this issue (CMIP5 vs CMIP6 in derived variables), could you please open (a) new issue(s) describing the specific problems with enough information? Sounds like even just loading them by themselves (ie outside of the amoc derivation) should be tackled first.

**Comment by ledm** [comment]
I've just noticed that the `basin` `axis_entry` also differs (slightly) between CMIP5 and CMIP6, which adds to my confusion.

CMIP5:
```
!============
axis_entry: basin
!============
!----------------------------------
! Axis attributes:
!----------------------------------
standard_name:    region
long_name:        ocean basin
!----------------------------------
! Additional axis information:
!----------------------------------
out_name:         region
type:             character
requested:        atlantic_arctic_ocean indian_pacific_ocean global_ocean        ! space-separated list of requested coordinates
must_have_bounds: no
coords_attrib:    region
!----------------------------------
```

CMIP6:
```
        "basin": {
            "standard_name": "region", 
            "units": "", 
            "axis": "", 
            "long_name": "ocean basin", 
            "climatology": "", 
            "formula": "", 
            "must_have_bounds": "no", 
            "out_name": "basin", 
            "positive": "", 
            "requested": [
                "atlantic_arctic_ocean", 
                "indian_pacific_ocean", 
                "global_ocean"
            ], 
            "requested_bounds": "", 
            "stored_direction": "", 
            "tolerance": "", 
            "type": "character", 
            "valid_max": "", 
            "valid_min": "", 
            "value": "", 
            "z_bounds_factors": "", 
            "z_factors": "", 
            "bounds_values": "", 
            "generic_level_name": ""
        }, 
```

The only real difference is the `out_name`; CMIP5: `region`, CMIP6: `basin`. For some reason, many of the CMIP6 model decided to use `sector` instead, which is in neither of these derivations. Who knows...

**Comment by ledm** [comment]
So, just received the news that several CMIP6 models use `msftmz` instead of `msftyz`. We knew this already, but I had put it aside to focus on the `msftyz` data.

In the AMOC derivation, the code is:

```
class DerivedVariable(DerivedVariableBase):
    """Derivation of variable `amoc`."""

    @staticmethod
    def required(project):
        """Declare the variables needed for derivation."""
        if project == 'CMIP5':
            required = [{'short_name': 'msftmyz'}]
        if project == 'CMIP6':
            required = [{'short_name': 'msftyz'}]
        return required
```
Is there a way to add a second `short_name` to the required data?

**Comment by zklaus** [comment]
Since the merging of #230 there is. Have a look at [the documentation](https://esmvaltool.readthedocs.io/projects/esmvalcore/en/latest/esmvalcore/derivation.html) and make use of the `optional` keyword.

**Comment by ledm** [comment]
thanks @zklaus, that solved it!

**Comment by helenelang** [comment]
1) I have updated the amoc.py:

    @staticmethod
    def required(project):
        """Declare the variables needed for derivation."""
        if project == 'CMIP5':
            required = [{'short_name': 'msftmyz'}]
        if project == 'CMIP6':
        #     required = [{'short_name': 'msftyz'}]
            required = [{'short_name': 'msftmyz', 'optional': True, },
                        {'short_name': 'msftyz', 'optional': True, },
                        {'short_name': 'msftmz', 'optional': True, }]

        return required

2) The recipe is updated (using either 'msftyz' or 'msftmz', depending on the model)
3) I still run into problems, please see the attached files.

[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/8669398/main_log_debug.txt)
[recipe_ocean_amoc.txt](https://github.com/ESMValGroup/ESMValCore/files/8669424/recipe_ocean_amoc.txt)

**Comment by valeriupredoi** [comment]
@helenelang let's not add to this old issue, but instead keep the discussion in https://github.com/ESMValGroup/ESMValTool/issues/2645 please
